cluster-autoscaler/cloudprovider/azure/azure_vms_pool.go

```diff
@@ -102,28 +120,260 @@ func (agentPool *VMsPool) MaxSize() int {
+		agentPool.clusterResourceGroup,
+		agentPool.clusterName,
+		agentPool.Name,
+		requestBody, nil)
+
+	if err != nil {
+		klog.Errorf("Failed to update agentpool %s in cluster %s with error: %v",
+			agentPool.azureRef.Name, agentPool.clusterName, err)
+		return err
+	}
+
+	updateCtx, cancel := getContextWithTimeout(updateRequestContextTimeout)
+	defer cancel()
+
+	if _, err = poller.PollUntilDone(updateCtx, nil); err == nil {
+		// success path
+		klog.Infof("agentPoolClient.BeginCreateOrUpdate for aks cluster %s agentpool %s succeeded", agentPool.clusterName, agentPool.Name)
+		agentPool.curSize = count
+		agentPool.lastSizeRefresh = time.Now()
+		klog.V(6).Infof("setVMsPoolNodeCount: invalidating cache, new size %d", count)
+		agentPool.manager.invalidateCache()
+		return nil
+	}
+
+	klog.Errorf("agentPoolClient.BeginCreateOrUpdate for aks cluster %s agentpool %s failed with error %s",
+		agentPool.clusterName, agentPool.Name, err)
+	return nil
 }
 
 // DeleteNodes extracts the providerIDs from the node spec and
-// delete or deallocate the nodes from the agent pool based on the scale down policy.
+// delete or deallocate the nodes based on the scale down policy of agentpool.
 func (agentPool *VMsPool) DeleteNodes(nodes []*apiv1.Node) error {
-	// TODO(wenxuan): Implement this method
-	return cloudprovider.ErrNotImplemented
+	currentSize, err := agentPool.getVMsPoolSize()
+	if err != nil {
+		return err
+	}
+
+	// if the target size is smaller than the min size, return an error
+	if int(currentSize) <= agentPool.MinSize() {
+		klog.V(3).Infof("min size %d reached, nodes will not be deleted", agentPool.MinSize())
+		return fmt.Errorf("min size %d reached, nodes will not be deleted", agentPool.MinSize())
+	}
+
+	var providerIDs []string
+	for _, node := range nodes {
+		belongs, err := agentPool.Belongs(node)
+		if err != nil {
+			return err
+		}
+		if !belongs {
+			return fmt.Errorf("node %s does not belong to agent pool %s", node.Name, agentPool.Name)
+		}
+
+		providerIDs = append(providerIDs, node.Spec.ProviderID)
+	}
+
+	return agentPool.scaleDownNodes(providerIDs)
+}
+
+// scaleDownNodes delete or deallocate the nodes from the agent pool based on the scale down policy.
+func (agentPool *VMsPool) scaleDownNodes(providerIDs []string) error {
+	agentPool.sizeMutex.Lock()
+	defer agentPool.sizeMutex.Unlock()
+
+	if len(providerIDs) == 0 {
+		return nil
+	}
+
+	klog.V(3).Infof("Deleting nodes from agent pool %s: %v", agentPool.Name, providerIDs)
+
+	deleteCtx, deleteCancel := getContextWithCancel()
+	defer deleteCancel()
+
+	machineNames := make([]*string, len(providerIDs))
+	for i, providerID := range providerIDs {
+		// extract the machine name from the providerID by splitting the providerID by '/' and get the last element
+		// The providerID look like this:
+		// "azure:///subscriptions/feb5b150-60fe-4441-be73-8c02a524f55a/resourceGroups/mc_wxrg_play-vms_eastus/providers/Microsoft.Compute/virtualMachines/aks-nodes-32301838-vms0"
```

Comment:
do we want to put real subs here since its open source? Given they can't do anything with the subscription, still probably good to not expose it.

```suggestion
		// "azure:///subscriptions/0000000-0000-0000-0000-00000000000/resourceGroups/mc_wxrg_play-vms_eastus/providers/Microsoft.Compute/virtualMachines/aks-nodes-32301838-vms0"
```

Reply:
Updated the sub as suggested.